codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -464,4 +403,166 @@ public Http2Settings pollSentSettings() {
     public Configuration configuration() {
         return frameWriter.configuration();
     }
+
+    /**
+     * Wrap a DATA frame so it can be written subject to flow-control. Note that this implementation assumes it
+     * only writes padding once for the entire payload as opposed to writing it once per-frame. This makes the
+     * {@link #size} calculation deterministic thereby greatly simplifying the implementation.
+     * <p>
+     * If frame-splitting is required to fit within max-frame-size and flow-control constraints we ensure that
+     * the passed promise is not completed until last frame write.
+     * </p>
+     */
+    private final class FlowControlledData extends FlowControlledBase {
+        private ByteBuf data;
+        private int size;
+
+        private FlowControlledData(ChannelHandlerContext ctx, Http2Stream stream, ByteBuf data, int padding,
+                                    boolean endOfStream, ChannelPromise promise) {
+            super(ctx, stream, padding, endOfStream, promise);
+            this.data = data;
+            size = data.readableBytes() + padding;
+        }
+
+        @Override
+        public int size() {
+            return size;
+        }
+
+        @Override
+        public void error(Throwable cause) {
+            ReferenceCountUtil.safeRelease(data);
```

Comment:
We're already protected against double-releasing by nulling out data on the first error & ReferenceCountUtil is safe with null.

If two writes fail only one of them gets to propagate to the promise, which is what really matters here. The interface of promise prevents us from aggregating these after we report the first one so we have to make a decision about which one to report and when. 

I favor reporting the first one to the promise and suppressing or logging the remaining ones largely because the first one is likely to be the most interesting and the subsequent ones are likely to be duplicates or derivatives of it.